src/Tools/dotnet-stack/Symbolicate.cs

```diff
@@ -0,0 +1,294 @@
+using System.Collections.Generic;
+using System.CommandLine;
+using System.CommandLine.Binding;
+using System.CommandLine.IO;
+using System.IO;
+using System.Linq;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Reflection.PortableExecutable;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.Diagnostics.Tools.Stack
+{
+    internal static class SymbolicateHandler
+    {
+        private static readonly Regex s_regex = new Regex(@" at (?<type>[\w+\.?]+)\.(?<method>\w+)\((?<params>.*)\) in (?<filename>[\w+\.?]+)(\.dll|\.ni\.dll): token (?<token>0x\d+)\+(?<offset>0x\d+)", RegexOptions.Compiled);
+        private static readonly Dictionary<string, MetadataReader> s_metadataReaderDictionary = new Dictionary<string, MetadataReader>();
+
+        delegate void SymbolicateDelegate(IConsole console, FileInfo inputPath, DirectoryInfo[] searchDir, FileInfo output, bool stdout);
+
+        /// <summary>
+        /// Get the line number from the Method Token and IL Offset at the stacktrace
+        /// </summary>
+        /// <param name="console"></param>
+        /// <param name="inputPath">The input path for file with stacktrace text</param>
+        /// <param name="searchDir">All paths in the directory to the assembly and pdb where the exception occurred</param>
+        /// <param name="output">The output path for the extracted line number data</param>
+        /// <returns></returns>
+        private static void Symbolicate(IConsole console, FileInfo inputPath, DirectoryInfo[] searchDir, FileInfo output, bool stdout)
+        {
+            try
+            {
+                if (output == null)
+                {
+                    output = new FileInfo(inputPath.FullName + ".symbolicated");
+                }
+
+                CreateSymbolicateFile(console, searchDir, inputPath.FullName, output.FullName, stdout);
+            }
+            catch (Exception e)
+            {
+                console.Error.WriteLine(e.Message);
+            }
+        }
+
+        private static void CreateSymbolicateFile(IConsole console, DirectoryInfo[] searchDir, string inputPath, string outputPath, bool isStdout)
+        {
+            try
+            {
+                SetMetadataReader(searchDir);
+
+                string ret = string.Empty;
+                using StreamWriter fileStreamWriter = new StreamWriter(new FileStream(outputPath, FileMode.Create, FileAccess.Write));
+                using StreamReader fileStreamReader = new StreamReader(new FileStream(inputPath, FileMode.Open, FileAccess.Read));
+                while (!fileStreamReader.EndOfStream)
+                {
+                    string line = fileStreamReader.ReadLine();
+                    if (!s_regex.Match(line).Success)
+                    {
+                        fileStreamWriter?.WriteLine(ret);
+                        if (isStdout) console.Out.WriteLine(ret);
+                        continue;
+                    }
```

Comment:
This section is duplicating work that happens in `TrySymbolicateLine`. `ret` from `TrySymbolicateLine` will already be the original input if it doesn't match the regex. This call to `RegularExpression.Match` can be removed.